03 - Data Science and Custom Skills/Azure Function Skills/SampleSkills/0_CustomSkillTemplate.cs

```diff
@@ -0,0 +1,171 @@
+            public List<OutputRecordWarnings> warnings { get; set; }
+        }
+
+        private class WebApiResponse
+        {
+            public WebApiResponse()
+            {
+                this.values = new List<OutputRecord>();
+            }
+
+            public List<OutputRecord> values { get; set; }
+        }
+        #endregion
+
+        [FunctionName("CustomSkillTemplate")]
+        public static async Task<IActionResult> Run(
+            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpRequest req,
+            ILogger log)
+        {
+            log.LogInformation("Custom skill: C# HTTP trigger function processed a request.");
+
+            // Read input, deserialize it and validate it.
+            var data = GetStructuredInput(req.Body);
+            if (data == null)
+            {
+                return new BadRequestObjectResult("The request schema does not match expected schema.");
+            }
+
+            // Calculate the response for each value.
+            var response = new WebApiResponse();
+            foreach (var record in data.values)
+            {
+                if (record == null || record.recordId == null) continue;
+
+                OutputRecord responseRecord = new OutputRecord();
+                responseRecord.recordId = record.recordId;
+
+                try
+                {
+                    responseRecord.data = DoWork(record.data).Result;
+                }
+                catch (Exception e)
+                {
+                    // Something bad happened, log the issue.
+                    var error = new OutputRecord.OutputRecordErrors
+                    {
+                        message = e.Message
+                    };
+
+                    responseRecord.errors = new List<OutputRecord.OutputRecordErrors>();
+                    responseRecord.errors.Add(error);
+                }
+                finally
+                {
+                    response.values.Add(responseRecord);
+                }
+            }
+
+            return new OkObjectResult(response);
+        }
+
+        private static WebApiRequest GetStructuredInput(Stream requestBody)
+        {
+            string request = new StreamReader(requestBody).ReadToEnd();
+
+            var data = JsonConvert.DeserializeObject<WebApiRequest>(request);
+            if (data == null)
+            {
+                return null;
+            }
+            return data;
+        }
+
+        /// <summary>
+        /// Replace this method with a method that enriches or transforms the data in
+        /// a meaningful manner
+        /// </summary>
+        /// <param name="inputField">Replace this with any fields you need to process.</param>
+        /// <returns>Feel free to change the return type to meet your needs. </returns>
+        async static Task<OutputRecord.OutputRecordData> DoWork(InputRecord.InputRecordData inputRecord)
```

Comment:
This doesn't need to be async for the purposes of the template, bit you can leave it as probably most scenarios do ultimately require that.  I would mark it as private static async though.